job-runner/chain/traverser.go

```diff
@@ -81,209 +110,169 @@ func (f *traverserFactory) Make(jobChain proto.JobChain) (Traverser, error) {
 	// for the chain: running, cleaning up, removing from repo when done, etc.
 	// And traverser and chain have the same lifespan: traverser is done when
 	// chain is done.
-	tr := NewTraverser(chain, f.chainRepo, f.rf, f.rmc)
+	cfg := TraverserConfig{
+		Chain:         chain,
+		ChainRepo:     f.chainRepo,
+		RunnerFactory: f.rf,
+		RMClient:      f.rmc,
+		ShutdownChan:  f.shutdownChan,
+		StopTimeout:   STOP_TIMEOUT,
+	}
+	tr := NewTraverser(cfg)
 	return tr, nil
 }
 
-// A traverser represents a job chain and everything needed to traverse it.
-type traverser struct {
-	// The chain that will be traversed.
-	chain *chain
-
-	// Repo for acessing/updating chains.
-	chainRepo Repo
-
-	// Factory for creating job runners.
-	rf runner.Factory
-
-	// Repo for keeping track of active jobs.
-	runnerRepo runner.Repo
-
-	// Used to stop a running traverser.
-	stopChan chan struct{}
-
-	// Queue for processing jobs that need to run.
-	runJobChan chan proto.Job
-
-	// Queue for processing jobs that are done running.
-	doneJobChan chan proto.Job
-
-	// Client for communicating with the Request Manager.
-	rmc rm.Client
-
-	// job.Id -> number of tries
-	jobTries map[string]uint
+// -------------------------------------------------------------------------- //
 
-	// mutex to access jobTries list
-	jtMux *sync.RWMutex
+type traverser struct {
+	reaperFactory ReaperFactory
+	reaper        JobReaper
+	runningJobs   map[string]struct{}
```

Comment:
I understand the usage but I think the runnerRepo needs to be the single source of truth for what's running.